### PR TITLE
Better handling of the app state and opened gecko sessions.

### LINKED ISSUE #91: Sessions are not correctly restored in some situations
After leaving the app, in case you do not "continue" or "quit" and just do something else, the sessions (opened private sessions) are not correctly restored when app is reopened.

### PR DESCRIPTION
**Pull request by keianhzo**
Fixes #91 

There were some situations were private sessions were not being correctly restored. For exmaple after leaving the app, in case you do not "continue" or "quit" and just go somewhere else an open the app later the app sessions were still alive but we were not restoring the private/public sessions correctly.

Please test in other headsets as I've only tested on Oculus Go and thetre might be some Android lifecycle related differences on those.

### CODE REVIEW

File: app/src/common/shared/org/mozilla/vrbrowser/VRBrowserActivity.java [hunk 1]
```diff
@@ -128,6 +129,21 @@ public void run() {
...
         initializeWorld();
+
+        int sessionId = SettingsStore.getInstance(this).getCurrentSessionId();
+        if (sessionId != SessionStore.NO_SESSION_ID) {
+            GeckoSession session = SessionStore.get().getSession(sessionId);
+            if (session != null) {
+                SessionStore.get().setCurrentSession(sessionId);
+                mPreviousSessionId = SettingsStore.getInstance(this).getPreviousSessionId();
+
+                if (session.getSettings().getBoolean(GeckoSessionSettings.USE_PRIVATE_MODE)) {
```
Comment (MortimerGoro) [inline app/src/common/shared/org/mozilla/vrbrowser/VRBrowserActivity.java]: The private browsing mode state setters are starting to get duplicated in different parts of the code. Could you use the SessionChangeListener.onCurrentSessionChange listener to handle the state and get rid of the setters in VRBrowserActivity?

File: app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java [hunk 2]
```diff
@@ -73,4 +73,25 @@ public void setTelemetryEnabled(boolean isEnabled) {
         TelemetryHolder.get().getConfiguration().setCollectionEnabled(isEnabled);
     }
 
+
+    public void setCurrentSeesionId(int sessionId) {
```
Comment (MortimerGoro) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java]: typo: seesion

File: app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java [hunk 3]
```diff
@@ -73,4 +73,25 @@ public void setTelemetryEnabled(boolean isEnabled) {
...
+        return mPrefs.getInt(mContext.getString(R.string.settings_key_session_id), SessionStore.NO_SESSION_ID);
+    }
+
+    public void setPreviousSeesionId(int sessionId) {
+        SharedPreferences.Editor editor = mPrefs.edit();
+        editor.putInt(mContext.getString(R.string.settings_key_previous_session_id), sessionId);
+        editor.commit();
+    }
+
+    public int getPreviousSessionId() {
```
Comment (MortimerGoro) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java]: Would it make sense to name the methods: getCurrentSessionId() and getCurrentPrivateSessionId()
I think it's clearer to call them that way if I have understand the usage correctly.

File: app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java [hunk 4]
```diff
@@ -73,4 +73,25 @@ public void setTelemetryEnabled(boolean isEnabled) {
...
+        SharedPreferences.Editor editor = mPrefs.edit();
+        editor.putInt(mContext.getString(R.string.settings_key_session_id), sessionId);
+        editor.commit();
+    }
+
+    public int getCurrentSessionId() {
+        return mPrefs.getInt(mContext.getString(R.string.settings_key_session_id), SessionStore.NO_SESSION_ID);
+    }
+
+    public void setPreviousSeesionId(int sessionId) {
```
Comment (MortimerGoro) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java]: The setter can also be automatically done using the SessionChangeListener.onCurrentSessionChange listener insteaf of using external calls

**Comment by MortimerGoro** [comment]
I think it would be a good idea to use the listeners to handle the state changes instead of the explicit setters. We are going to add more ways to change sessions in the future (carrousel, voice input) so it would be better to handle the state in a single place

File: app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java (hunk 3 above)
Comment (MortimerGoro) [reply app/src/common/shared/org/mozilla/vrbrowser/ui/SettingsStore.java]: mmm you'd also need a extra variable to know if the current is public/private. So it's ok

**Comment by keianhzo** [comment]
Yes, using SessionChangeListener is the right way, changed all but the current/previous session id setters as when the activityis detroyed we don't get any session callback but we still need to save the session for restoring when the activity is recreated (as sessions are still alive in Gecko).

File: app/src/common/shared/org/mozilla/vrbrowser/ui/TopBarWidget.java [hunk 5]
```diff
@@ -56,6 +59,8 @@ public void onClick(View view) {
         });
 
         mAudio = AudioEngine.fromContext(aContext);
+
+        SessionStore.get().addSessionChangeListener(this);
```
Comment (MortimerGoro) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/TopBarWidget.java]: unregister the listener on
```
 @Override
public void releaseWidget() {
}
```

Comment (keianhzo) [reply app/src/common/shared/org/mozilla/vrbrowser/ui/TopBarWidget.java]: Fixed

**Pull request merged**